<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">
 
  <!-- Camera Sensor -->
  <xacro:macro name="camera_sensor" params="parent xyz rpy">
    <link name="camera_link">
      <visual>
        <geometry>
          <box size="0.3 0.2 0.3"/>
        </geometry>
        <material name="black">
          <color rgba="0 0 0 1"/>
        </material>
      </visual>
      <inertial>
        <mass value="0.2"/>
        <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001"/>
      </inertial>
    </link>
 
    <joint name="camera_joint" type="fixed">
      <parent link="${parent}"/>
      <child link="camera_link"/>
      <origin xyz="${xyz}" rpy="${rpy}"/>
    </joint>
 
    <!-- Gazebo Camera Plugin -->
    <gazebo reference="camera_link">
      <sensor type="camera" name="camera_sensor">
        <update_rate>30.0</update_rate>
        <camera name="camera">
          <horizontal_fov>1.047</horizontal_fov>
          <image>
            <width>640</width>
            <height>480</height>
            <format>R8G8B8</format>
          </image>
          <clip>
            <near>0.1</near>
            <far>100</far>
          </clip>
        </camera>
        <plugin name="camera_controller" filename="libgazebo_ros_openni_kinect.so">
          <alwaysOn>true</alwaysOn>
          <updateRate>30.0</updateRate>
          <cameraName>camera</cameraName>
          <imageTopicName>color/image_raw2</imageTopicName>
          <cameraInfoTopicName>color/camera_info2</cameraInfoTopicName>
          <depthImageTopicName>depth/image_raw2</depthImageTopicName>
          <depthCameraInfoTopicName>depth/camera_info2</depthCameraInfoTopicName>
          <frameName>camera_link</frameName>
          <hackBaseline>0.07</hackBaseline>
          <distortionK1>0.0</distortionK1>
          <distortionK2>0.0</distortionK2>
          <distortionK3>0.0</distortionK3>
          <distortionT1>0.0</distortionT1>
          <distortionT2>0.0</distortionT2>
        </plugin>
      </sensor>
    </gazebo>
  </xacro:macro>
 
</robot>